### PR TITLE
feat(bitbucket-server): support bearer token authentication

### LINKED ISSUE #14900: Add Authorization type of "Bearer <token>" for Bitbucket Server
### What would you like Renovate to be able to do?

I would like renovate to be able to connect to bitbucket-server using just a token, similar to some of the other platforms, I want to be able to provide a token and use the Bearer Authorisation strategy. 

### If you have any ideas on how this should be implemented, please tell us here.

I think [The Auth check](https://github.com/renovatebot/renovate/blob/7d13826854f2dcef8996f59347defbdc8164990d/lib/modules/platform/bitbucket-server/index.ts#L91) could be modified to check if username + password OR token. 

This isn't implemented already is it? If so please let me know if I can make this work. 

Bitbucket recommend using the auth token without username - see below link

"For project or repository tokens, we recommend only using Bearer Auth without your username:"
https://confluence.atlassian.com/bitbucketserver/http-access-tokens-939515499.html

Any advice appreciated! 
Thanks

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by jaecktec**
Add support for project/repository http access token

## Changes

allowed token authentication by passing the authentication to the rest api and simple-git as extra-header

## Context

fixes https://github.com/renovatebot/renovate/issues/14900 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/bitbucket-server/index.ts [hunk 1]
```diff
@@ -302,7 +316,9 @@ export async function getPrList(refreshCache?: boolean): Promise<Pr[]> {
     };
     if (!config.ignorePrAuthor) {
       searchParams['role.1'] = 'AUTHOR';
-      searchParams['username.1'] = config.username;
+      if(config.username){
```
Comment (rarkins) [inline lib/modules/platform/bitbucket-server/index.ts:319]: Can the username be "learned" using the API? It's possible in other platforms to do a "whoami" type query to the API

File: lib/util/git/config.ts [hunk 2]
```diff
@@ -17,13 +17,16 @@ export function getNoVerify(): GitNoVerifyOption[] {
   return noVerify;
 }
 
-export function simpleGitConfig(): Partial<SimpleGitOptions> {
+export function simpleGitConfig(extraConfig: SimpleGitOptions['config'] = []): Partial<SimpleGitOptions> {
```
Comment (rarkins) [inline lib/util/git/config.ts:20]: Isn't it possible to pass token in the same way other platforms do?

File: lib/modules/platform/bitbucket-server/index.ts [hunk 3]
```diff
@@ -205,6 +218,7 @@ export async function initRepo({
       url,
       cloneSubmodules,
       fullClone: true,
+      authorization: opts.token,
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:221]: no, will already be handled via hostrule and is already added to opts.

you need to update the `getRepoGitUrl` function above to handle token auth.

File: lib/util/git/config.ts (hunk 2 above)
Comment (viceice) [reply lib/util/git/config.ts:20]: yes, revert

File: lib/util/git/index.ts [hunk 4]
```diff
@@ -232,7 +232,7 @@ export async function initRepo(args: StorageConfig): Promise<void> {
   config.ignoredAuthors = [];
   config.additionalBranches = [];
   config.branchIsModified = {};
-  git = simpleGit(GlobalConfig.get('localDir'), simpleGitConfig()).env({
+  git = simpleGit(GlobalConfig.get('localDir'), simpleGitConfig(config.authorization ? [`http.extraHeader=Authorization: Bearer ${config.authorization}`] : undefined)).env({
```
Comment (viceice) [inline lib/util/git/index.ts:235]: revert

File: lib/util/git/types.ts [hunk 5]
```diff
@@ -17,6 +17,7 @@ export type LongCommitSha = string & { __longCommitSha: never };
 export interface StorageConfig {
   currentBranch?: string;
   url: string;
+  authorization?: string,
```
Comment (viceice) [inline lib/util/git/types.ts:20]: revert.

File: lib/util/git/config.ts [hunk 6]
```diff
@@ -17,13 +17,16 @@ export function getNoVerify(): GitNoVerifyOption[] {
...
   const config: Partial<SimpleGitOptions> = {
     completion: {
       onClose: true,
       onExit: false,
     },
-    config: ['core.quotePath=false'],
+    config: [
+      ...extraConfig,
+      'core.quotePath=false',
+    ],
```
Comment (viceice) [inline lib/util/git/config.ts:29]: revert

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/platform/bitbucket-server/index.ts (hunk 1 above)
Comment (jaecktec) [reply lib/modules/platform/bitbucket-server/index.ts:319]: no, afaik the point is to be anonymous. the tokens are for a repository and not for a user. They are no PAT.

Comment (viceice) [reply lib/modules/platform/bitbucket-server/index.ts:319]: please check what's the author for PRs created by such a token.

**Comment by simschla** [comment]
I'm confused: is this PR closed because the feature will not be added or is it not possible?

File: lib/util/git/config.ts (hunk 2 above)
Comment (jaecktec) [reply lib/util/git/config.ts:20]: the token can't be passed as an URL argument

**Comment by jaecktec** [comment]
I can't find the suggested places required to be changed.

**Pull request closed without merging**